src/main/java/wooteco/subway/dao/line/JdbcLineDao.java

```diff
@@ -0,0 +1,72 @@
+package wooteco.subway.dao.line;
+
+import lombok.RequiredArgsConstructor;
+import org.springframework.jdbc.core.JdbcTemplate;
+import org.springframework.jdbc.core.RowMapper;
+import org.springframework.jdbc.support.GeneratedKeyHolder;
+import org.springframework.jdbc.support.KeyHolder;
+import org.springframework.stereotype.Repository;
+import wooteco.subway.domain.Line;
+
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.util.List;
+import java.util.Optional;
+
+@RequiredArgsConstructor
+@Repository
+public class JdbcLineDao implements LineDao {
+
+    private final JdbcTemplate jdbcTemplate;
+
+    @Override
+    public Line save(Line line) {
+        String sql = "INSERT INTO line (name, color) VALUES (?, ?)";
+        KeyHolder keyHolder = new GeneratedKeyHolder();
+        jdbcTemplate.update(con -> {
+            PreparedStatement preparedStatement = con.prepareStatement(sql, new String[]{"id"});
+            preparedStatement.setString(1, line.getName());
+            preparedStatement.setString(2, line.getColor());
+            return preparedStatement;
+        }, keyHolder);
+        return Line.of(keyHolder.getKey().longValue(), line.getName(), line.getColor(), line.getSections());
+    }
+
+    @Override
+    public Optional<Line> findByNameAndColor(String name, String color) {
+        String sql = "SELECT * FROM line WHERE (name = ? OR color = ?)";
+        return jdbcTemplate.query(sql, getRowMapper(), name, color).stream().findAny();
+    }
+
+    @Override
+    public Optional<Line> findById(Long lineId) {
+        String sql = "SELECT * FROM line WHERE id = ?";
+        return jdbcTemplate.query(sql, getRowMapper(), lineId).stream().findAny();
+    }
```

Comment:
```suggestion
    public Optional<Line> findById(Long lineId) {
        String sql = "SELECT * FROM line WHERE id = ?";

        try {
            return Optional.ofNullable(jdbcTemplate.queryForObject(sql, getRowMapper(), lineId));
        } catch (EmptyResultDataAccessException e) {
            return Optional.empty();
        }
    }
```

`queryForObject()`로 수정하면 동작이 안 된다고 하셨는데, 이렇게 하면 되지 않나요? 어떻게 수정하셨던 건지 궁금해요.

Reply:
혹시 저 경우에는 try~catch 에 감싸야해서 가독성이 떨어져 보이는데, query 에서 찾는 방식은 비용 문제 때문에 좋지 않은 걸까요? 혹은 유니크 제약 조건에 대해서도 생각해봤는데 그 부분은 두 부분 모두 문제될 것 같아서 이것은 이유가 아닐 것 같네요.

Reply:
비용 문제는 아니고, https://github.com/woowacourse/atdd-subway-map/pull/63#discussion_r629498103 와 같은 상황을 생각했을 때 `findAny()`는 위험해 보여서요. 단 하나의 row만 리턴되어야 하는 상황이라는 것이 명시적으로 표현되기도 하고요.